[feat] Update some functions for issue #36

dacbinh opened the pull request:
Added 4 functions for UD operation on regimens and regimen_details table each include role check and basic error handling include:
update_regimen -> void: this function update regimens table and regimen_details table
delete_regimen -> void: this function soft delete a regimen record
remove_medicine_from_regimen -> void: this function remove a mediccine from table regimen_details
add_medicine_into_regimen -> void: this function add a medicine into regimen_detaills, included check for invalid medicine_id and regimen_id

dacbinh committed:
issue #36
added 4 functions for UD operation on regimens and regimen_details table
each include role check and basic error handling

maik205 commented on fns/remove_medicine_from_regimen.sql:
```diff
@@ -0,0 +1,17 @@
+returns void as 
+$$
+begin
+    IF (auth.uid() IS NULL) OR NOT ((SELECT roles FROM user_details WHERE user_id = auth.uid()) @> ARRAY['administrator', 'manager']::role[]) THEN
```
use `check_roles(roles role[])`

maik205 commented on fns/remove_medicine_from_regimen.sql:
```diff
@@ -0,0 +1,17 @@
+        where regimen_id = p_regimen_id and medicine_id = p_medicine_id;
+    begin
+    exception
+    when others then
```
No need

maik205 commented on fns/delete_regimen.sql:
```diff
@@ -0,0 +1,19 @@
+returns void as
+$$
+begin
+    if(auth.uid() is null) or not ((select roles from user_details where user_id = auth.uid()) @> array['administrator', 'manager']::role[]) 
```
use `check_roles(roles role[])`

maik205 commented on fns/delete_regimen.sql:
```diff
@@ -0,0 +1,19 @@
+    where regimen_id = p_regimen_id;
+    exception
+    when others then
+    raise exception 'Something went wrong while trying to delete regimen: %', SQLERRM;
```
No need

maik205 commented on fns/update_regimen.sql:
```diff
@@ -0,0 +1,54 @@
+    override_json json;
+begin
+    --check user role
+    if(auth.uid() is null) or not ((select roles from user_details where user_id = auth.uid()) @> array['administrator', 'manager']::role[])
```
use `check_roles(roles role[])`

maik205 commented on fns/update_regimen.sql:
```diff
@@ -0,0 +1,54 @@
+    --update the table regimens
+        update regimens
+        set
+            name = p_name,
```
name should also be optional

maik205 commented on fns/update_regimen.sql:
```diff
@@ -0,0 +1,54 @@
+            );
+        end loop;
+        exception
+        when others then
```
No need

maik205 commented on fns/add_medicine_into_regimen.sql:
```diff
@@ -0,0 +1,51 @@
+$$
+begin
+--check user role
+    if(auth.uid() is null) or not ((select roles from user_details where user_id = auth.uid()) @> array['administrator', 'manager']::role[])
```
use `check_roles(roles role[])` or RLS policy to prohibit INSERTS. 

Any INSERT will cause an error.

Any SELECT will return an empty result.

Any UPDATE or DELETE will be successful, but affect no row. This is because these operations have to read (scan) the table before they modify data, and that scan will return no rows like in the SELECT case. Since no rows are affected by the data modification, no error is thrown.

maik205 commented on fns/add_medicine_into_regimen.sql:
```diff
@@ -0,0 +1,51 @@
+        );
+        --in case something else went wrong
+        exception
+        when others then
```
No need, create confusion and unhelpful.

maik205 commented on fns/add_medicine_into_regimen.sql:
```diff
@@ -0,0 +1,51 @@
+    raise exception 'Regimen id does not exist: %s', p_regimen_id;
+    end if;
+--start adding medicine to regimen
+    begin
```
Why start another begin block?

dacbinh commented:
Thank you for your review, changes will come in the next update.

dacbinh committed:
issue #36
removed some unneccesary code
added a function check_roles()
name in regimen is now optional too.

maik205 commented on fns/update_regimen.sql:
```diff
@@ -0,0 +1,55 @@
+            is_for_adults = coalesce(p_is_for_adults, is_for_adults)       
+        where regimen_id = p_regimen_id;
+        --delete the medicine that need to be override
+        for override_json in select * from json_array_elements(p_override_medicines)
```
Please also do checks for json validity.

maik205 replied:
json validity should function specific

dacbinh replied:
One more question, should this function return a boolean value or directly raise an exception?

maik205 commented:
The function already exists on public schema and raises exceptions.

On Tue, 10 Jun 2025 at 12:46 dakbink ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In fns/update_regimen.sql
> <https://github.com/DorayakiDevs/haivy-db/pull/37#discussion_r2136959774>:
>
> > +    -- end if;
> +    if not(check_roles(array['administrator', 'manager'])::role[]) then
> +    then
> +    raise exception 'You do not have permission to perform this action';
> +    end if;
> +    begin
> +    --update the table regimens
> +        update regimens
> +        set
> +            name = coalesce(p_name, name),
> +            description = coalesce(p_description, description),
> +            level = coalesce(p_level, level),
> +            is_for_adults = coalesce(p_is_for_adults, is_for_adults)
> +        where regimen_id = p_regimen_id;
> +        --delete the medicine that need to be override
> +        for override_json in select * from json_array_elements(p_override_medicines)
>
> One more question, should this function return a boolean value or directly
> raise an exception?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/DorayakiDevs/haivy-db/pull/37#discussion_r2136959774>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AIKWPNDGECXG5P3MEOOVZ6T3CZWMFAVCNFSM6AAAAAB62X2HGWVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDSMJSGAZTKMBSGY>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

dacbinh commented:
Please leave the fucntion name here so i can search for it because i am currently , thanks.

maik205 commented:
![image](https://github.com/user-attachments/assets/54ddfd05-4a4c-49b6-9f18-fbb312aeec21)

maik205 commented:
It does not exist, please create it yourself. Thanks

On Tue, 10 Jun 2025 at 12:54 dakbink ***@***.***> wrote:

> *dacbinh* left a comment (DorayakiDevs/haivy-db#37)
> <https://github.com/DorayakiDevs/haivy-db/pull/37#issuecomment-2957767561>
>
> Sorry for misleading, are we talking about check_roles() or the function
> to validate json format.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/DorayakiDevs/haivy-db/pull/37#issuecomment-2957767561>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AIKWPNCMOC3TQ25IGZZT7WD3CZXKFAVCNFSM6AAAAAB62X2HGWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDSNJXG43DONJWGE>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

dacbinh committed:
issue #36
added function for json validation (daily_dosage_schedule)

dacbinh committed:
issue #36
Removed uneccesary check_roles() -> this function already exist in public schema

dacbinh committed:
issue #36
replaced old json validation with new one for functions in this list

dacbinh committed:
issue#36
replaced some old functions role check with new ones

dacbinh commented on fns/add_medicine_into_regimen.sql:
```diff
@@ -0,0 +1,43 @@
+--eheck for valid json
+    select validate_daily_dosage_schedule(p_daily_dosage_schedule);
+--check if medicine exist
+    if not exists (select 1 from mecidicines where id = p_medicine_id)
```
Thanks for reminding me, it will be change in future updates

dacbinh committed:
issue #36
minor: fixed some typo
